flashbax/vault/vault.py

```diff
@@ -0,0 +1,448 @@
+        if base_path_exists:
+            print(f"Loading vault found at {self._base_path}")
+            # Vault exists, so we load the metadata to access the structure etc.
+            self._metadata = json.loads(metadata_path.read_text())
+
+            # Ensure minor versions match
+            assert (self._metadata["version"] // 1) == (VERSION // 1)
+
+        elif experience_structure is not None:
+            print(f"New vault created at {self._base_path}")
+
+            # Create the necessary dirs for the vault
+            os.makedirs(self._base_path)
+
+            # Ensure provided metadata is json serialisable
+            metadata_json_ready = jax.tree_util.tree_map(
+                lambda obj: str(obj)
+                if not isinstance(obj, (bool, str, int, float, type(None)))
+                else obj,
+                metadata,
+            )
+
+            # We save the structure of the buffer state
+            #  e.g. [(128, 100, 4), jnp.int32]
+            # We will use this structure to map over the data stores later
+            serialised_experience_structure = jax.tree_map(
+                lambda x: [str(x.shape), x.dtype.name],
+                serialize_tree(
+                    # Get shape and dtype of each leaf, without serialising the data itself
+                    jax.eval_shape(
+                        lambda: experience_structure,
+                    ),
+                ),
+            )
+
+            # Construct metadata
+            self._metadata = {
+                "version": VERSION,
+                "structure": serialised_experience_structure,
+                **(metadata_json_ready or {}),  # Allow user to save extra metadata
+            }
+            # Dump metadata to file
+            metadata_path.write_text(json.dumps(self._metadata))
+        else:
+            # If the vault does not exist already, and no experience_structure is provided to create
+            # a new vault, we cannot proceed.
+            raise ValueError(
+                "Vault does not exist and no experience_structure was provided."
+            )
+
+        # We must now build the tree structure from the metadata, whether the metadata was created
+        #  here or loaded from file
+        if experience_structure is None:
+            # Since the experience structure is not provided, we simply use the metadata as is.
+            #  The result will always be a dictionary.
+            self._tree_structure = self._metadata["structure"]
+        else:
+            # If experience structure is provided, we try deserialise into that structure
+            self._tree_structure = deserialize_tree(
+                self._metadata["structure"],
+                target=experience_structure,
+            )
+
+        # Each leaf of the fbx_state.experience maps to a data store, so we tree map over the
+        #  tree structure to create each of the data stores.
+        self._all_ds = jax.tree_util.tree_map_with_path(
```

Comment:
I'd just name this `self.datastore` as it seems to be the main one

Reply:
happy to rename, but it should be something like `self._all_datastores` because each leaf has its own datastore, and this tree holds all of them :)